src/execution/StepExecutionManager.ts

```diff
@@ -43,27 +62,79 @@ export class StepExecutionManager {
     const existingProcess = step.execution.process.find(
       (p) => p.type === currentProcessType
     )
+
     // Check token approval only if fromToken is not the native token => no approval needed in that case
-    if (
+
+    const checkForAllowance =
       !existingProcess?.txHash &&
-      !isZeroAddress(step.action.fromToken.address)
-    ) {
-      await checkAllowance(
+      !isZeroAddress(step.action.fromToken.address) &&
+      (shouldBatchTransactions || !isMultisigSigner)
+
+    if (checkForAllowance) {
+      const populatedTransaction = await checkAllowance(
         signer,
         step,
         statusManager,
         settings,
         fromChain,
-        this.allowUserInteraction
+        this.allowUserInteraction,
+        shouldBatchTransactions
       )
+
+      if (populatedTransaction) {
+        const { to, data } = populatedTransaction
+
+        if (to && data) {
+          // allowance doesn't need value
+          const cleanedPopulatedTransaction: BaseTransaction = {
+            value: '',
+            to,
+            data,
+          }
+
+          multisigBatchTransactions.push(cleanedPopulatedTransaction)
+        }
+      }
     }
 
     // STEP 2: Get transaction
     let process = statusManager.findOrCreateProcess(step, currentProcessType)
 
     if (process.status !== 'DONE') {
+      const multisigProcess = step.execution.process.find(
+        (p) => !!p.multisigTxHash
+      )
+
       try {
-        let transaction: TransactionResponse
+        if (isMultisigSigner && multisigProcess) {
+          if (!multisigProcess) {
+            throw new ValidationError('Multisig process is undefined.')
+          }
+          if (!config.multisigConfig?.getMultisigTransactionDetails) {
+            throw new ValidationError(
+              '"getMultisigTransactionDetails()" is missing in Multisig config.'
+            )
+          }
+
+          const multisigTxHash = multisigProcess.multisigTxHash
+
+          if (!multisigTxHash) {
+            // need to check what happens in failed tx
+            throw new ValidationError(
+              'Multisig internal transaction hash is undefined.'
+            )
+          }
+
+          await updateSafeRouteProcess(
```

Comment:
Abstracted the logic of fetching status from SAFE and processing response to be used when the transaction is initiated as well.

In the `fetchMultisigStatus` function the recursion is present, hence we need to trigger this to initiate the polling.


And currently the polling logic is present only upon resuming the route (here).

Abstracting the logic for this and reusing it post initiation: https://github.com/lifinance/sdk/pull/150/files#diff-bc44d69ee7c901547808f63d12468d9523a358d036d8f69db430969460cec5e1R321